framework/include/base/MooseApp.h

```diff
@@ -178,13 +178,20 @@ class MooseApp : public ConsoleStreamInterface, public libMesh::ParallelObject
   void setOutputFileBase(const std::string & output_file_base)
   {
     _output_file_base = output_file_base;
+    _file_base_set_by_user = true;
   }
 
   /**
-   * Override the selection of the output file base name.
+   * Get the output file base name.
    */
   std::string getOutputFileBase() const;
 
+  /**
+   * Whether or not setOutputFileBase is called or this app has file_base
+   * in the Outputs input block.
+   */
+  bool outputFileBaseSetByUser() const { return _file_base_set_by_user; }
```

Comment:
It feels like this is the wrong API. I know this isn't your fault, but the encapsulation around the whole output naming scheme is horribly broken. I can see that there are already several unrelated objects making decisions about file_names, but this PR is adding to the mess. You are adding even more places where these decisions are being made.

Let's take a step back. It's clear that the Output system can't completely make this decision on it's own as we require information about where we are at in a Multiapp tree to make some of these decisions. It seems like your motivation for starting this cleanup is good. You wanted to have a single place to obtain this information and maybe the App is the right place for that (at least for starters). However, by choosing an API where you have to expose something about how the name got created you started heading down the wrong path. Now objects that don't have anything to do with the filenames have to know how to name things and have to check this odd API to make those decisions. I can see two or three new places where you are calling this and then branching to create the desired name. All of that logic should be ideally encapsulated, right?